### PR TITLE
Allow metal3-dev-env to run on a generic ubuntu VM

### PR DESCRIPTION
**Pull request by elfosardo**
Currently metal3-dev-env is configured to run on a preconfigured ubuntu host.
Some configuration parameters and options are given as assumed, but to run it
in a generic VM ( based on ubuntu focal or jammy), we need to update the
configuration to take into consideration a default clean OS configuration.

### CODE REVIEW

**Comment by elfosardo** [comment]
/test-centos-integration-release-1-3
" [FAILED] The e2e test config file is not valid "

File: vm-setup/roles/firewall/defaults/main.yml
Comment (tuminoid) [inline vm-setup/roles/firewall/defaults/main.yml:6]: Could you document `PROVISIONING_NETWORK` in `vars.md` while at it? Seems to be missing.

File: vm-setup/roles/firewall/defaults/main.yml
Comment (tuminoid) [inline vm-setup/roles/firewall/defaults/main.yml:4]: `CLUSTER_PROVISIONING_INTERFACE` with default value of `ironicendpoint` is defined in `vars.md`, we should not hardcode it here?

**Review by tuminoid — COMMENTED** [review]
Two small things related to environment variables.

Also, please squash before merge.

**Review by furkatgofurov7 — COMMENTED** [review]
Only one nit, not sure if CI failures related to this PR though

**Comment by furkatgofurov7** [comment]
> Also, please squash before merge.

+1

**Comment by tuminoid** [comment]
I verified the changes on Ubuntu 22.04 and Ubuntu 20.04, but when testing Ubuntu 18.04 I found out that we have neglected both Ubuntu 18.04 and Centos 8 when we have uplifted Ansible. Both of those have Python 3.6 so the newest Ansible version supporting 3.6 is 4.10.0. I've created PR to address that shortcoming here: https://github.com/metal3-io/metal3-dev-env/pull/1164 and with those fixes, I managed to verify this PR on 18.04 as well.

File: 01_prepare_host.sh
Comment (elfosardo) [reply 01_prepare_host.sh:33]: mmmm not sure this can be done without complicating the code too much, we would save one line but probably add more, we need to disable the kvm module first, and only if we detect virtualization support (either intel or amd)

**Comment by elfosardo** [comment]
squashed and rebased

File: vm-setup/roles/firewall/defaults/main.yml
Comment (elfosardo) [reply vm-setup/roles/firewall/defaults/main.yml]: I just remembered why we put this here, it's note exported, not sure if it's a bug, didn't have the time to investigate, but without this here ansible  will complain

Comment (tuminoid) [reply vm-setup/roles/firewall/defaults/main.yml:4]: I did not mean to remove it, but take it from env with a fallback, like the other variables around it. Like:
`cluster_provisioning_interface: "{{ lookup('env', 'CLUSTER_PROVISIONING_INTERFACE') | default('ironicendpoint', true) }}" `

**Comment by elfosardo** [comment]
weird, this works locally
I'm seeing errors on ipa related to the certificate, maybe we overlooked something?

**Comment by mboukhalfa** [comment]
> weird, this works locally I'm seeing errors on ipa related to the certificate, maybe we overlooked something?

Well locally with same image that the CI use I have got the kcp machine provsioned but worker machine pending forever ! I did not get time to go deep with this issue

File: vm-setup/roles/firewall/defaults/main.yml
Comment (elfosardo) [reply vm-setup/roles/firewall/defaults/main.yml:4]: oh yeah, my bad, will change that

**Comment by mboukhalfa** [comment]
> > weird, this works locally I'm seeing errors on ipa related to the certificate, maybe we overlooked something?
> 
> Well locally with same image that the CI uses I have got the kcp machine provsioned but worker machine pending forever ! I did not get time to go deep with this issue

it seems that calico is not running on the kcp machine:
```
$ kubectl get pods -A                                                                                                                                                                                      
NAMESPACE     NAME                                       READY   STATUS     RESTARTS        AGE                                                                                                            
kube-system   calico-kube-controllers-5f94594857-77l4z   0/1     Pending    0               4m14s                                                                                                          
kube-system   calico-node-wqlpv                          0/1     Init:1/3   2               4m14s                                                                                                          
kube-system   etcd-test1-bclst                           1/1     Running    0               9m22s                                                                                                          
kube-system   kube-apiserver-test1-bclst                 1/1     Running    1               9m46s                                                                                                          
kube-system   kube-controller-manager-test1-bclst        1/1     Running    2               8m43s                                                                                                          
kube-system   kube-scheduler-test1-bclst                 1/1     Running    1 (2m32s ago)   9m27s
```

File: vm-setup/roles/firewall/tasks/iptables.yaml
Comment (mboukhalfa) [inline vm-setup/roles/firewall/tasks/iptables.yaml:63]: should not this be 
```suggestion
    in_interface: "{{ provisioning_interface }}"
```

File: vm-setup/roles/firewall/tasks/iptables.yaml
Comment (mboukhalfa) [inline vm-setup/roles/firewall/tasks/iptables.yaml:54]: what is difference between `provisioning_interface` and `cluster_provisioning_interface`

File: vm-setup/roles/libvirt/tasks/vm_setup_tasks.yml
Comment (mboukhalfa) [inline vm-setup/roles/libvirt/tasks/vm_setup_tasks.yml:19]: why defaulting qemu inside a VM ? we used KVM as default ! can explain the impact of this when using generic ubuntu VM ?

**Review by mboukhalfa — COMMENTED** [review]
some remarks

File: vm-setup/roles/libvirt/tasks/vm_setup_tasks.yml
Comment (elfosardo) [reply vm-setup/roles/libvirt/tasks/vm_setup_tasks.yml:19]: pure kvm won't always work in a VM, nested virtualization needs to be enabled as well as hardware pass-through
for example, in the ironic upstream CI that is openstack based, we use qemu on the VMs because of security concerns of some of the public cloud providers where the CI is configured

File: vm-setup/roles/firewall/tasks/iptables.yaml
Comment (elfosardo) [reply vm-setup/roles/firewall/tasks/iptables.yaml:54]: the provisioning_interface is a direct reference to the provisioning network, while the cluster_provisioning_interface refers to the ironic endpoint

File: vm-setup/roles/libvirt/tasks/vm_setup_tasks.yml
Comment (mboukhalfa) [reply vm-setup/roles/libvirt/tasks/vm_setup_tasks.yml:19]: Could qemu demand more resources ?

Comment (elfosardo) [reply vm-setup/roles/libvirt/tasks/vm_setup_tasks.yml:19]: not really, but I haven't seen the same in ironic upstream CI
to make this pass I'll introduce a variable for the kvm/qemu switch so we can set that when there is no kvm support and make it work fine in ironic CI
thanks for the help!

Comment (elfosardo) [reply vm-setup/roles/libvirt/tasks/vm_setup_tasks.yml:19]: I'm leaving this autoset option here for future cases or if we decide to set LIBVIRT_DOMAIN_TYPE to null

**Comment by mboukhalfa** [comment]
> /test-ubuntu-e2e-integration-main

the tests passed but the deletion of the cluster failed ! might a flake ! can we remove the debuging commits like increasing ansible verbosity not sure if you would like to keep adding 443 squash commits

File: vm-setup/roles/libvirt/tasks/vm_setup_tasks.yml
Comment (mboukhalfa) [reply vm-setup/roles/libvirt/tasks/vm_setup_tasks.yml:19]: When using kvm the nested virtualization can be enabled from the host or from the infrastructure that virtualize the host ?

Comment (elfosardo) [reply vm-setup/roles/libvirt/tasks/vm_setup_tasks.yml:19]: nested virtualization is enabled on the host, therefore whoever manages the infrastructure where the hosts run can decide to enable it or not

File: lib/common.sh
Comment (elfosardo) [reply lib/common.sh:347]: I will add that, thanks
in the ironic CI the qemu configuration is different based on infra, last one I checked and verified is 8CPUs/8GB

**Review by dtantsur — APPROVED** [review]

**Comment by mboukhalfa** [comment]
/assign  @kashifest 
cc @furkatgofurov7 @lentzi90

**Review by kashifest — APPROVED** [review]

**Comment by kashifest** [comment]
Giving @furkatgofurov7 some time to review

**Review by tuminoid — APPROVED** [review]
Thanks for the PR! Looking solid to me.
/lgtm

**Comment by kashifest** [comment]
/hold cancel
I think theres enough review on this PR now

**Comment by furkatgofurov7** [comment]
> Giving @furkatgofurov7 some time to review

This looks good overall to me, but I am not sure how this affects BML lab CI (mostly the firewall changes). Unfortunately it looks like the tests are broken now, I don't think we can verify it.

/lgtm

**Pull request merged**
